### PR TITLE
AT hooks

### PR DESCRIPTION
**Pull request by eivindj-nordic**

### CODE REVIEW

File: doc/nrf/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -376,6 +376,16 @@ Bootloader libraries
 Modem libraries
 ---------------
 
+* Added the :ref:`at_cmd_hook_readme` library to add AT commands hooks with application callbacks.
```
Comment (lemrey) [inline doc/nrf/releases/release-notes-changelog.rst:379]: suggestion: `.. to add application callbacks for AT commands`

File: lib/at_cmd_hook/at_cmd_hook.c [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#include <stdio.h>
+#include <zephyr/toolchain/common.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(at_cmd_hook, CONFIG_AT_CMD_HOOK_LOG_LEVEL);
+
+static int at_cmd_hook_sys_init(const struct device *unused)
```
Comment (lemrey) [inline lib/at_cmd_hook/at_cmd_hook.c:20]: note: signature is changing with next upmerge

File: lib/at_cmd_hook/at_cmd_hook.c [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	int err;
+	extern struct nrf_modem_at_cmd_hook _nrf_modem_at_cmd_hook_list_start[];
+	extern struct nrf_modem_at_cmd_hook _nrf_modem_at_cmd_hook_list_end[];
+
+#define HOOK_COUNT (_nrf_modem_at_cmd_hook_list_end - _nrf_modem_at_cmd_hook_list_start)
+
+	err = nrf_modem_at_cmd_hooks_set(_nrf_modem_at_cmd_hook_list_start, HOOK_COUNT,
+					 CONFIG_AT_HOOK_POST_CB_LEN_MAX);
+	LOG_INF("AT hooks enabled with %d entries", HOOK_COUNT);
+	printk("AT hooks enabled with %d entries\n", HOOK_COUNT);
```
Comment (lemrey) [inline lib/at_cmd_hook/at_cmd_hook.c:33]: Is `printk` leftover?

Comment (eivindj-nordic) [reply lib/at_cmd_hook/at_cmd_hook.c:33]: yes

File: lib/at_cmd_hook/at_cmd_hook.c [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+LOG_MODULE_REGISTER(at_cmd_hook, CONFIG_AT_CMD_HOOK_LOG_LEVEL);
+
+static int at_cmd_hook_sys_init(void)
+{
+	int err;
+	extern struct nrf_modem_at_cmd_hook _nrf_modem_at_cmd_hook_list_start[];
+	extern struct nrf_modem_at_cmd_hook _nrf_modem_at_cmd_hook_list_end[];
+
+#define HOOK_COUNT (_nrf_modem_at_cmd_hook_list_end - _nrf_modem_at_cmd_hook_list_start)
```
Comment (gregersrygg) [inline lib/at_cmd_hook/at_cmd_hook.c:26]: Does the define need to be inside the function?

**Review by gregersrygg — APPROVED** [review]

File: doc/nrf/libraries/modem/at_cmd_hook.rst [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+If the AT command string is longer than :kconfig:option:`CONFIG_AT_HOOK_POST_CB_LEN_MAX`, it will be truncated in the post callback.
+
+Adding an AT command hook
+=========================
+
+The application can define an AT command hook to receive a callback using the :c:macro:`AT_CMD_HOOK` macro.
+An AT command hook has a name, a filter string, two callback functions, and a state.
+Multiple parts of the application can define their own hooks. If there are multiple filters matching the AT command, all matching callbacks will be triggered.
+
+The following code snippet shows how to add a custom command that triggers on ``+MYCOMMAND`` calls:
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply doc/nrf/libraries/modem/at_cmd_hook.rst:43]: Correct, I will change this to use CFUN, as that is probably the one that will be used the most.

**Pull request merged**